{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d68b42e2-7eb2-45ce-8cea-e15444d9154b",
   "metadata": {},
   "source": [
    "# Workshop on Statistical Genetics and Genetic Epidemiology STAGE-Quebec\n",
    "## Theme 2 - Molecular Phenotypes in Genetic Epidemiology\n",
    "\n",
    "By Marc-André Legault (Université de Montréal) and Qihuang Zhang (McGill University)\n",
    "\n",
    "**July 31 - August 1, 2025**\n",
    "\n",
    "### Introduction\n",
    "\n",
    "In this workshop, we will conduct a transcriptome-wide association analysis (TWAS) to identify type 2 diabetes (T2D) genes. More precisely, we will test the association between predicted gene expression in various tissues and T2D. We will use real summary association statistics from a recent meta-analysis involving 180,834 cases with T2D and over a million controls [Mahajan _et al._ (2022)](https://www.nature.com/articles/s41588-022-01058-3). We will use both the [S-PrediXcan](https://www.nature.com/articles/s41467-018-03621-1) and [TWAS FUSION](https://www.nature.com/articles/ng.3506) to conduct the TWAS analyses. The GWAS summary statistics have been pre-processed to ensure they are on the same genome build (GRCh38) as the gene expression prediction models distributed by these software. We will also focus on a subset of tissues from GTEx to reduce the computational burden.\n",
    "\n",
    "We split the workshop contents into three sections.\n",
    "\n",
    "1. S-PrediXcan (this Notebook)\n",
    "2. TWAS / FUSION [(link)](2a-FUSION.ipynb)\n",
    "3. Comparing both approaches [(link)](3a-compare_TWAS_models.ipynb)\n",
    "\n",
    "There is a file browser on the left pane that can be used to browse the workshop contents.\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "    <p>\n",
    "    The <strong>local</strong> folder is shared with your computer. Everything else will not be saved if you close the notebook server! Make sure that everything that should persist is store in the local folder.\n",
    "    </p>\n",
    "    <p>\n",
    "    The contents of this folder are stored in your home directory (e.g. C:\\Users\\[your username]\\STAGE2025_workshop_theme2 or <br /> /Users/[your username]/STAGE2025_workshop_theme2).\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "We can now begin by better understanding S-PrediXcan and running it using our pre-processed GWAS data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffb1f99e-87c3-47b8-b422-e99013a8cb28",
   "metadata": {},
   "source": [
    "### Exploring the GWAS summary statistics\n",
    "\n",
    "The GWAS summary statistics have been harmonized with [PredictDB](https://predictdb.hakyimlab.org/) models which are the gene expression predictions models distributed as companion data to S-PrediXcan. The data format is the same for every tissue:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "73589759-60a8-45ff-81fe-3d4773cd758a",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking at file ../data/PrediXcan/gwas_harmonized/harmonized_en_Whole_Blood.tsv.gz\n",
      "Column numbers and titles:\n",
      "     1\tchromosome\n",
      "     2\tbase_pair_location\n",
      "     3\teffect_allele\n",
      "     4\tother_allele\n",
      "     5\tbeta\n",
      "     6\tstandard_error\n",
      "     7\teffect_allele_frequency\n",
      "     8\tneg_log_10_p_value\n",
      "     9\trsid_x\n",
      "    10\tvariant_id\n",
      "    11\tpredictdb_id\n",
      "    12\trsid_y\n",
      "\n",
      "File excerpt:\n",
      "chromosome\tbase_pair_location\teffect_allele\tother_allele\tbeta\tstandard_error\trsid_x\n",
      "chr1\t785910\tC\tG\t-0.0176\t0.0262\trs12565286\n",
      "chr1\t817186\tA\tG\t0.0048\t0.0098\trs3094315\n",
      "chr1\t817341\tG\tA\t0.0045\t0.0097\trs3131972\n",
      "chr1\t825532\tT\tC\t0.0069\t0.0099\trs1048488\n",
      "chr1\t825767\tC\tT\t0.0066\t0.0098\trs3115850\n",
      "chr1\t833068\tA\tG\t0.0\t0.0117\trs12562034\n",
      "chr1\t841742\tT\tA\t0.0078\t0.0101\trs2980319\n",
      "chr1\t843942\tG\tA\t-0.007\t0.0101\trs4040617\n",
      "chr1\t849670\tA\tG\t0.0063\t0.0098\trs2905062\n"
     ]
    }
   ],
   "source": [
    "filename=\"../data/PrediXcan/gwas_harmonized/harmonized_en_Whole_Blood.tsv.gz\"\n",
    "echo \"Looking at file $filename\"\n",
    "echo \"Column numbers and titles:\"\n",
    "echo \"$(zcat $filename | head -n 1 | tr '\\t' '\\n' | nl)\"\n",
    "echo\n",
    "echo \"File excerpt:\"\n",
    "zcat $filename | head -n 10 | cut -f 1-6,9"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d84d2d6b-1a38-4f2f-b2d2-2f538b04af58",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "    Interpret the displayed columns.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61ab83c3-8a39-410c-869a-d1d1eeeef27f",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Gene expression prediction models\n",
    "\n",
    "Now, we can look at the gene expression prediction models for the GTEx tissue \"Whole blood\". In PredictDB, the models are distributed as sqlite3 databases. The databases have two tables: ``extra`` and ``weights`` with the following structure:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cad73eb5-cada-4838-bec1-44a6bddaef07",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CREATE TABLE `extra` (\n",
      "  `gene` TEXT,\n",
      "  `genename` TEXT,\n",
      "  `gene_type` TEXT,\n",
      "  `alpha` REAL,\n",
      "  `n_snps_in_window` INTEGER,\n",
      "  `n.snps.in.model` INTEGER,\n",
      "  `test_R2_avg` REAL,\n",
      "  `test_R2_sd` REAL,\n",
      "  `cv_R2_avg` REAL,\n",
      "  `cv_R2_sd` REAL,\n",
      "  `in_sample_R2` REAL,\n",
      "  `nested_cv_fisher_pval` REAL,\n",
      "  `nested_cv_converged` INTEGER,\n",
      "  `rho_avg` REAL,\n",
      "  `rho_se` REAL,\n",
      "  `rho_zscore` REAL,\n",
      "  `pred.perf.R2` REAL,\n",
      "  `pred.perf.pval` REAL,\n",
      "  `pred.perf.qval` REAL,\n",
      "  `phi` REAL\n",
      ");\n",
      "CREATE INDEX metab_model_summary ON extra (gene);\n"
     ]
    }
   ],
   "source": [
    "# Schema for the \"extra\" table:\n",
    "sqlite3 ../data/PrediXcan/predictdb/en_Whole_Blood.db \".schema extra\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22f8bc0c-b16d-45f8-93b7-099483098945",
   "metadata": {},
   "source": [
    "``extra`` contains metadata on the predictive model for every gene. You will find the gene name, the Ensembl gene ID, the number of SNPs that were included in the model along with statistics on the predictive performance of the gene expression model. \n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "Comment on the importance of the <code>pred.perf.pval</code> column. What is it for?\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9ab2df2-b8c9-4ba3-9538-0d7fce689f83",
   "metadata": {},
   "source": [
    "The weights table contains the information on the gene expression model. For every ``gene`` (Ensembl gene ID), it will list the variants that are used to predict gene expression (i.e. eQTLs) along with their prediction ``weight``. As a concrete example, we can extract the model used to predict the expression for ENSG00000157823.16 (Adaptor Related Protein Complex 3 Subunit Sigma 2, _AP3S2_). For brevity, we will sort the variants by absolute weight, and only show the top 10 contributors to the prediction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "33627747-e995-4c01-b12c-889ec39b0ab6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene                rsid        varID                   ref_allele  eff_allele  weight             \n",
      "------------------  ----------  ----------------------  ----------  ----------  -------------------\n",
      "ENSG00000157823.16  rs4363847   chr15_90547112_T_C_b38  T           C           0.128329668848438  \n",
      "ENSG00000157823.16  rs12898828  chr15_89817721_A_T_b38  A           T           -0.109067828981916 \n",
      "ENSG00000157823.16  rs11073814  chr15_88835653_A_T_b38  A           T           0.0524825659916266 \n",
      "ENSG00000157823.16  rs3853641   chr15_89900547_G_C_b38  G           C           -0.0467714907373544\n",
      "ENSG00000157823.16  rs16943673  chr15_89864201_T_C_b38  T           C           -0.0326710086433488\n",
      "ENSG00000157823.16  rs10520684  chr15_89881770_C_A_b38  C           A           -0.0295110487238457\n",
      "ENSG00000157823.16  rs1879530   chr15_88902553_G_A_b38  G           A           0.0263159897074107 \n",
      "ENSG00000157823.16  rs2165069   chr15_89863887_C_T_b38  C           T           -0.0251527944185157\n",
      "ENSG00000157823.16  rs4932250   chr15_89818473_G_A_b38  G           A           -0.0242119510907331\n",
      "ENSG00000157823.16  rs3759831   chr15_89871410_C_G_b38  C           G           -0.0223217371169919\n"
     ]
    }
   ],
   "source": [
    "sqlite3 ../data/PrediXcan/predictdb/en_Whole_Blood.db <<EOF\n",
    ".mode columns\n",
    ".header on\n",
    "select * from weights\n",
    "  where gene='ENSG00000157823.16'\n",
    "  order by abs(weight) desc limit 10;\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fb3cdb7-5b2e-415f-9d07-9ff0fe4985f7",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "Why are some weights positive and some negative? Can you mention a way of estimating the prediction weights?\n",
    "</div>\n",
    "\n",
    "### TWAS analysis using S-PrediXcan\n",
    "\n",
    "Now that we have explored the structure of PredictDB gene expression prediction models and that we have GWAS summary statistics, we can implement the TWAS analysis using S-PrediXcan. The code below may run for a few minutes. If it takes too long, you can stop it using the stop button at the top of this window, and use the checkpoint files available in the repository.\n",
    "\n",
    "The call to SPrediXcan takes 3 input files:\n",
    "\n",
    "- model_db_path: This contains the weights to predict the expression of the genes included in the TWAS. It is the sqlite3 database provided by PredictDB.\n",
    "- covariance: This file contains LD estimates between the SNPs. It accounts for the covariance between genetic variants.\n",
    "- gwas_file: This is the GWAS summary statistics files containing the association statistics between genetic variants and T2D. We previously harmonized the raw GWAS results to make sure that the variants were consistent with the ones available in PredictDB.\n",
    "\n",
    "The parameters with the ``_column`` suffix are used so that the software knows how to interpret the GWAS summary statistics file. You can convince yourself that the names provided below are appropriate by looking back at the excerpt from the GWAS summary statistics file above.\n",
    "\n",
    "Finally, we provide an estimate of the SNP heritability ($h^2$) for our phenotype and the GWAS sample size (N). These parameters allow the software to recalibrate the association statistics to account for polygenicity, as described in [Liang _et al. medRxiv_ 2024](https://www.biorxiv.org/content/10.1101/2023.10.17.562831v2). In short, polygenic traits have many causal variants with weak effect sizes scattered across the genome. Many of these variants will be in LD with the eQTLs included in the gene expression prediction models, which could lead to spurious associations. Providing these statistics allows S-PrediXcan to recalibrate the TWAS association statistics to account for this effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98d19b80-4a11-47a9-b3a9-9c7b5b96c477",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Running S-PrediXcan for tissue Adipose_Subcutaneous\n",
      "\n",
      "INFO - Processing GWAS command line parameters\n",
      "INFO - Building beta for ../data/PrediXcan/gwas_harmonized/harmonized_en_Adipose_Subcutaneous.tsv.gz and ../data/PrediXcan/predictdb/en_Adipose_Subcutaneous.db\n",
      "INFO - Reading input gwas with special handling: ../data/PrediXcan/gwas_harmonized/harmonized_en_Adipose_Subcutaneous.tsv.gz\n",
      "INFO - Processing input gwas\n",
      "INFO - Calculating zscore from se and beta\n",
      "INFO - Aligning GWAS to models\n",
      "INFO - Trimming output\n",
      "INFO - Successfully parsed input gwas in 2.0491462999925716 seconds\n",
      "INFO - Started metaxcan process\n",
      "INFO - Loading model from: ../data/PrediXcan/predictdb/en_Adipose_Subcutaneous.db\n",
      "INFO - Loading covariance data from: ../data/PrediXcan/predictdb/en_Adipose_Subcutaneous.txt.gz\n",
      "INFO - Processing loaded gwas\n",
      "INFO - Started metaxcan association\n",
      "INFO - 10 % of model's snps found so far in the gwas study\n",
      "INFO - 20 % of model's snps found so far in the gwas study\n",
      "INFO - 30 % of model's snps found so far in the gwas study\n",
      "INFO - 40 % of model's snps found so far in the gwas study\n",
      "INFO - 50 % of model's snps found so far in the gwas study\n",
      "INFO - 60 % of model's snps found so far in the gwas study\n",
      "INFO - 70 % of model's snps found so far in the gwas study\n",
      "INFO - 80 % of model's snps found so far in the gwas study\n",
      "INFO - 86 % of model's snps used\n",
      "INFO - Calibrating pvalue and zscore using phi in model, N and h2\n",
      "INFO - The pvalue and zscore have been calibrated successfully\n",
      "INFO - Sucessfully processed metaxcan association in 31.825913532011327 seconds\n",
      "\n",
      "\n",
      "Running S-PrediXcan for tissue Artery_Coronary\n",
      "\n",
      "INFO - Processing GWAS command line parameters\n",
      "INFO - Building beta for ../data/PrediXcan/gwas_harmonized/harmonized_en_Artery_Coronary.tsv.gz and ../data/PrediXcan/predictdb/en_Artery_Coronary.db\n",
      "INFO - Reading input gwas with special handling: ../data/PrediXcan/gwas_harmonized/harmonized_en_Artery_Coronary.tsv.gz\n",
      "INFO - Processing input gwas\n",
      "INFO - Calculating zscore from se and beta\n",
      "INFO - Aligning GWAS to models\n",
      "INFO - Trimming output\n",
      "INFO - Successfully parsed input gwas in 0.9646924749977188 seconds\n",
      "INFO - Started metaxcan process\n",
      "INFO - Loading model from: ../data/PrediXcan/predictdb/en_Artery_Coronary.db\n",
      "INFO - Loading covariance data from: ../data/PrediXcan/predictdb/en_Artery_Coronary.txt.gz\n",
      "INFO - Processing loaded gwas\n",
      "INFO - Started metaxcan association\n",
      "INFO - 10 % of model's snps found so far in the gwas study\n",
      "INFO - 20 % of model's snps found so far in the gwas study\n",
      "INFO - 30 % of model's snps found so far in the gwas study\n",
      "INFO - 40 % of model's snps found so far in the gwas study\n",
      "INFO - 50 % of model's snps found so far in the gwas study\n",
      "INFO - 60 % of model's snps found so far in the gwas study\n",
      "INFO - 70 % of model's snps found so far in the gwas study\n",
      "INFO - 80 % of model's snps found so far in the gwas study\n",
      "INFO - 86 % of model's snps used\n",
      "INFO - Calibrating pvalue and zscore using phi in model, N and h2\n",
      "INFO - The pvalue and zscore have been calibrated successfully\n",
      "INFO - Sucessfully processed metaxcan association in 16.167929881004966 seconds\n",
      "\n",
      "\n",
      "Running S-PrediXcan for tissue Brain_Cortex\n",
      "\n",
      "INFO - Processing GWAS command line parameters\n",
      "INFO - Building beta for ../data/PrediXcan/gwas_harmonized/harmonized_en_Brain_Cortex.tsv.gz and ../data/PrediXcan/predictdb/en_Brain_Cortex.db\n",
      "INFO - Reading input gwas with special handling: ../data/PrediXcan/gwas_harmonized/harmonized_en_Brain_Cortex.tsv.gz\n",
      "INFO - Processing input gwas\n",
      "INFO - Calculating zscore from se and beta\n",
      "INFO - Aligning GWAS to models\n",
      "INFO - Trimming output\n",
      "INFO - Successfully parsed input gwas in 1.6318275509984232 seconds\n",
      "INFO - Started metaxcan process\n",
      "INFO - Loading model from: ../data/PrediXcan/predictdb/en_Brain_Cortex.db\n",
      "INFO - Loading covariance data from: ../data/PrediXcan/predictdb/en_Brain_Cortex.txt.gz\n",
      "INFO - Processing loaded gwas\n",
      "INFO - Started metaxcan association\n",
      "INFO - 10 % of model's snps found so far in the gwas study\n",
      "INFO - 20 % of model's snps found so far in the gwas study\n",
      "INFO - 30 % of model's snps found so far in the gwas study\n",
      "INFO - 40 % of model's snps found so far in the gwas study\n",
      "INFO - 50 % of model's snps found so far in the gwas study\n",
      "INFO - 60 % of model's snps found so far in the gwas study\n",
      "INFO - 70 % of model's snps found so far in the gwas study\n",
      "INFO - 80 % of model's snps found so far in the gwas study\n",
      "INFO - 86 % of model's snps used\n",
      "INFO - Calibrating pvalue and zscore using phi in model, N and h2\n",
      "INFO - The pvalue and zscore have been calibrated successfully\n",
      "INFO - Sucessfully processed metaxcan association in 24.970216764995712 seconds\n",
      "\n",
      "\n",
      "Running S-PrediXcan for tissue Liver\n",
      "\n",
      "INFO - Processing GWAS command line parameters\n",
      "INFO - Building beta for ../data/PrediXcan/gwas_harmonized/harmonized_en_Liver.tsv.gz and ../data/PrediXcan/predictdb/en_Liver.db\n",
      "INFO - Reading input gwas with special handling: ../data/PrediXcan/gwas_harmonized/harmonized_en_Liver.tsv.gz\n",
      "INFO - Processing input gwas\n",
      "INFO - Calculating zscore from se and beta\n",
      "INFO - Aligning GWAS to models\n",
      "INFO - Trimming output\n",
      "INFO - Successfully parsed input gwas in 0.8767220279987669 seconds\n",
      "INFO - Started metaxcan process\n",
      "INFO - Loading model from: ../data/PrediXcan/predictdb/en_Liver.db\n",
      "INFO - Loading covariance data from: ../data/PrediXcan/predictdb/en_Liver.txt.gz\n",
      "INFO - Processing loaded gwas\n",
      "INFO - Started metaxcan association\n",
      "INFO - 10 % of model's snps found so far in the gwas study\n",
      "INFO - 20 % of model's snps found so far in the gwas study\n",
      "INFO - 30 % of model's snps found so far in the gwas study\n",
      "INFO - 40 % of model's snps found so far in the gwas study\n",
      "INFO - 50 % of model's snps found so far in the gwas study\n",
      "INFO - 60 % of model's snps found so far in the gwas study\n",
      "INFO - 70 % of model's snps found so far in the gwas study\n",
      "INFO - 80 % of model's snps found so far in the gwas study\n",
      "INFO - 86 % of model's snps used\n",
      "INFO - Calibrating pvalue and zscore using phi in model, N and h2\n",
      "INFO - The pvalue and zscore have been calibrated successfully\n",
      "INFO - Sucessfully processed metaxcan association in 14.002740316995187 seconds\n",
      "\n",
      "\n",
      "Running S-PrediXcan for tissue Muscle_Skeletal\n",
      "\n",
      "INFO - Processing GWAS command line parameters\n",
      "INFO - Building beta for ../data/PrediXcan/gwas_harmonized/harmonized_en_Muscle_Skeletal.tsv.gz and ../data/PrediXcan/predictdb/en_Muscle_Skeletal.db\n",
      "INFO - Reading input gwas with special handling: ../data/PrediXcan/gwas_harmonized/harmonized_en_Muscle_Skeletal.tsv.gz\n",
      "INFO - Processing input gwas\n",
      "INFO - Calculating zscore from se and beta\n",
      "INFO - Aligning GWAS to models\n",
      "INFO - Trimming output\n",
      "INFO - Successfully parsed input gwas in 1.5832355199963786 seconds\n",
      "INFO - Started metaxcan process\n",
      "INFO - Loading model from: ../data/PrediXcan/predictdb/en_Muscle_Skeletal.db\n",
      "INFO - Loading covariance data from: ../data/PrediXcan/predictdb/en_Muscle_Skeletal.txt.gz\n",
      "INFO - Processing loaded gwas\n",
      "INFO - Started metaxcan association\n",
      "INFO - 10 % of model's snps found so far in the gwas study\n",
      "INFO - 20 % of model's snps found so far in the gwas study\n",
      "INFO - 30 % of model's snps found so far in the gwas study\n",
      "INFO - 40 % of model's snps found so far in the gwas study\n",
      "INFO - 50 % of model's snps found so far in the gwas study\n",
      "INFO - 60 % of model's snps found so far in the gwas study\n",
      "INFO - 70 % of model's snps found so far in the gwas study\n",
      "INFO - 80 % of model's snps found so far in the gwas study\n",
      "INFO - 86 % of model's snps used\n",
      "INFO - Calibrating pvalue and zscore using phi in model, N and h2\n",
      "INFO - The pvalue and zscore have been calibrated successfully\n",
      "INFO - Sucessfully processed metaxcan association in 24.109790089001763 seconds\n",
      "\n",
      "\n",
      "Running S-PrediXcan for tissue Pancreas\n",
      "\n",
      "INFO - Processing GWAS command line parameters\n",
      "INFO - Building beta for ../data/PrediXcan/gwas_harmonized/harmonized_en_Pancreas.tsv.gz and ../data/PrediXcan/predictdb/en_Pancreas.db\n",
      "INFO - Reading input gwas with special handling: ../data/PrediXcan/gwas_harmonized/harmonized_en_Pancreas.tsv.gz\n",
      "INFO - Processing input gwas\n",
      "INFO - Calculating zscore from se and beta\n",
      "INFO - Aligning GWAS to models\n",
      "INFO - Trimming output\n",
      "INFO - Successfully parsed input gwas in 1.3673999450111296 seconds\n",
      "INFO - Started metaxcan process\n",
      "INFO - Loading model from: ../data/PrediXcan/predictdb/en_Pancreas.db\n",
      "INFO - Loading covariance data from: ../data/PrediXcan/predictdb/en_Pancreas.txt.gz\n",
      "INFO - Processing loaded gwas\n",
      "INFO - Started metaxcan association\n",
      "INFO - 10 % of model's snps found so far in the gwas study\n",
      "INFO - 20 % of model's snps found so far in the gwas study\n",
      "INFO - 30 % of model's snps found so far in the gwas study\n",
      "INFO - 40 % of model's snps found so far in the gwas study\n",
      "INFO - 50 % of model's snps found so far in the gwas study\n",
      "INFO - 60 % of model's snps found so far in the gwas study\n",
      "INFO - 70 % of model's snps found so far in the gwas study\n",
      "INFO - 80 % of model's snps found so far in the gwas study\n",
      "INFO - 86 % of model's snps used\n",
      "INFO - Calibrating pvalue and zscore using phi in model, N and h2\n",
      "INFO - The pvalue and zscore have been calibrated successfully\n",
      "INFO - Sucessfully processed metaxcan association in 22.07476048899116 seconds\n",
      "\n",
      "\n",
      "Running S-PrediXcan for tissue Whole_Blood\n",
      "\n",
      "INFO - Processing GWAS command line parameters\n",
      "INFO - Building beta for ../data/PrediXcan/gwas_harmonized/harmonized_en_Whole_Blood.tsv.gz and ../data/PrediXcan/predictdb/en_Whole_Blood.db\n",
      "INFO - Reading input gwas with special handling: ../data/PrediXcan/gwas_harmonized/harmonized_en_Whole_Blood.tsv.gz\n",
      "INFO - Processing input gwas\n",
      "INFO - Calculating zscore from se and beta\n",
      "INFO - Aligning GWAS to models\n",
      "INFO - Trimming output\n",
      "INFO - Successfully parsed input gwas in 1.772096025000792 seconds\n",
      "INFO - Started metaxcan process\n",
      "INFO - Loading model from: ../data/PrediXcan/predictdb/en_Whole_Blood.db\n",
      "INFO - Loading covariance data from: ../data/PrediXcan/predictdb/en_Whole_Blood.txt.gz\n",
      "INFO - Processing loaded gwas\n",
      "INFO - Started metaxcan association\n",
      "INFO - 10 % of model's snps found so far in the gwas study\n",
      "INFO - 20 % of model's snps found so far in the gwas study\n",
      "INFO - 30 % of model's snps found so far in the gwas study\n",
      "INFO - 40 % of model's snps found so far in the gwas study\n",
      "INFO - 50 % of model's snps found so far in the gwas study\n",
      "INFO - 60 % of model's snps found so far in the gwas study\n",
      "INFO - 70 % of model's snps found so far in the gwas study\n",
      "INFO - 80 % of model's snps found so far in the gwas study\n",
      "INFO - 86 % of model's snps used\n",
      "INFO - Calibrating pvalue and zscore using phi in model, N and h2\n",
      "INFO - The pvalue and zscore have been calibrated successfully\n",
      "INFO - Sucessfully processed metaxcan association in 23.84649238199927 seconds\n"
     ]
    }
   ],
   "source": [
    "tissues=\"Adipose_Subcutaneous Artery_Coronary Brain_Cortex Liver Muscle_Skeletal Pancreas Whole_Blood\"\n",
    "\n",
    "for tissue in $tissues; do\n",
    "    echo\n",
    "    echo\n",
    "    echo \"Running S-PrediXcan for tissue ${tissue}\"\n",
    "    echo\n",
    "    SPrediXcan.py \\\n",
    "        --model_db_path ../data/PrediXcan/predictdb/en_${tissue}.db \\\n",
    "        --covariance ../data/PrediXcan/predictdb/en_${tissue}.txt.gz \\\n",
    "        --gwas_file ../data/PrediXcan/gwas_harmonized/harmonized_en_${tissue}.tsv.gz \\\n",
    "        --snp_column rsid_x \\\n",
    "        --effect_allele_column effect_allele \\\n",
    "        --non_effect_allele_column other_allele \\\n",
    "        --output_file /workshop/local/results/twas_en_${tissue}.csv \\\n",
    "        --beta_column beta \\\n",
    "        --se_column standard_error \\\n",
    "        --gwas_h2 0.15 \\\n",
    "        --gwas_N 1000000\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c1c674e-66a9-4e88-94da-abd303e08dda",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "    You can open one of the results files in the ``results`` directory to see the TWAS results for yourself.\n",
    "</div>\n",
    "\n",
    "In the [next notebook](1b-S-PrediXcan-interpretation.ipynb), we will interpret the TWAS results we generated using the R language."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
